### ISSUE TITLE
[1.15.1-1.15.2] OBJ Blocks rendering mostly dark on some faces and bright on others

### ISSUE THREAD

**Issue by StanCEmpire** [MinecraftForge/MinecraftForge#6425]
So I've been playing around with the forge OBJ loader and I've ran into this same issue with all my blocks where the blocks are rendering very dark on most faces but then brighter on other faces. On the forge discord I asked about this and I was told that it could be a problem with the normals on my model. However after going back into both my OBJ file and my 3D software and confirming that the normals were being written properly and that my normals were facing the right direction (Both were confirmed by members of the discord server). One member of the discord server figured it was most likely a bug with the OBJLoader. I could be wrong but I figured it was best to post this here anyway just to ensure.

I've attached some images of the dark blocks below. Now of course these are test blocks and I would not use an OBJ model just for a default cube. These are merely for reference purposes but this issue has persisted across all of myOBJ  models recently. I'm assuming it is mostly to do with the render changes in 1.15.

Errors On Startup: 0

Using forge-1.15.1-30.0.35
![1](https://user-images.githubusercontent.com/59638278/72190039-a0535580-3452-11ea-84cb-3d56ffb7ad68.png)
![2](https://user-images.githubusercontent.com/59638278/72190041-a0ebec00-3452-11ea-95f9-587ee1f2af23.png)

Model JSON
```json {

	"parent": "forge:block/default",
	"loader": "forge:obj",
	"model": "stancempiresextrabuildingblocks:models/block/cobblestone_panel.obj",
	"flip-v": true,
	"textures":
	{
		"particle": "block/cobblestone"
	}

}
```
OBJ File
```obj
# Blender v2.81 (sub 16) OBJ File: 'PlaceHolderBlock.blend'
# www.blender.org
mtllib cobblestone_panel.mtl
o Block_Cube.001
v 0.000000 0.000000 1.000000
v 0.000000 1.000000 1.000000
v 0.000000 0.000000 0.000000
v 0.000000 1.000000 0.000000
v 1.000000 0.000000 1.000000
v 1.000000 1.000000 1.000000
v 1.000000 0.000000 0.000000
v 1.000000 1.000000 0.000000
vt 0.500000 0.749957
vt 0.250043 0.749957
vt 0.250043 0.500000
vt 0.500000 0.500000
vt 0.250043 0.250043
vt 0.500000 0.250043
vt 0.250043 0.000087
vt 0.500000 0.000087
vt 0.500000 0.999913
vt 0.250043 0.999913
vt 0.749957 0.500000
vt 0.749957 0.749957
vt 0.000087 0.500000
vt 0.000087 0.749957
vn -1.0000 0.0000 0.0000
vn 0.0000 0.0000 -1.0000
vn 1.0000 0.0000 0.0000
vn 0.0000 0.0000 1.0000
vn 0.0000 -1.0000 0.0000
vn 0.0000 1.0000 0.0000
usemtl PlaceHolderMAT
s off
f 1/1/1 2/2/1 4/3/1 3/4/1
f 3/4/2 4/3/2 8/5/2 7/6/2
f 7/6/3 8/5/3 6/7/3 5/8/3
f 5/9/4 6/10/4 2/2/4 1/1/4
f 3/4/5 7/11/5 5/12/5 1/1/5
f 8/13/6 4/3/6 2/2/6 6/14/6
```
MTL File
```mtl
# Blender MTL File: 'PlaceHolderBlock.blend'
# Material Count: 1

newmtl PlaceHolderMAT
Ns 323.999994
Ka 1.000000 1.000000 1.000000
Kd 0.800000 0.800000 0.800000
Ks 0.500000 0.500000 0.500000
Ke 0.000000 0.000000 0.000000
Ni 1.000000
d 1.000000
illum 2
map_Kd stancempiresextrabuildingblocks:block/cobblestone_panel```

**Comment by gigaherz** [comment]
Hi, could you retest in forge 30.0.39? I am not certain if your issue was caused by the same problem I just fixed or not.

**Comment by StanCEmpire** [comment]
Well in 30.0.39 the blocks are working for the most part. There are a few instances however where there are some odd shadow effects on the block but for the most part the blocks are the correct brightness. I've included images of both the old version (30.0.35) in comparison to the newer version which for the most part works perfectly. Ive included some images to demonstrate the minor artefacts appearing on some blocks in specific positions.

OLD VERSION
![Old](https://user-images.githubusercontent.com/59638278/72315271-4a96db80-36e6-11ea-8852-af4880761e52.png)

NEW VERSION (including some blocks with a few odd artefacts)
![New](https://user-images.githubusercontent.com/59638278/72315290-57b3ca80-36e6-11ea-8ac5-96760f559abd.png)
![New2](https://user-images.githubusercontent.com/59638278/72315291-584c6100-36e6-11ea-9909-b65dbeb61ddb.png)
![New3](https://user-images.githubusercontent.com/59638278/72315292-584c6100-36e6-11ea-923b-e81c17dbc651.png)

So in summary I would say it works very well for the most part 😃  thanks for the speedy update too. Just a few minor shadows on a few blocks but mostly very good 👍 

Just as a side note some of the weird shadows could be due to low in-game lighting but I'm unsure as there are some cases where blocks are well lit but still have some strange shadows

**Comment by StanCEmpire** [comment]
Ok so i did a bit of testing and I think that the weird shadows (that I mentioned in the comment above) are only weird because they are appearing on the wrong side of the block. In the image above (in the previous comment) you can see on the bottom row of blocks there is a dark shadow on it. That shadow is normally on the bottom side of the block.

![issue](https://user-images.githubusercontent.com/59638278/72586057-731c1100-3944-11ea-8f2e-76ac9840902f.png)
As seen here on the stone block, the faint shadow is suppose to be on the lower half of the block

This shadow only seems to occur if there are blocks underneath the custom OBJ block but is not affected by blocks above it.

I have tested this on the latest forge version (30.0.41) and the issue still persists

**Comment by gigaherz** [comment]
Heya, I was tracking down the issue last night, and I did in fact figure out what the problem is: the minecraft lighting engine **does care** about what the specific ordering of the vertices is. This means like, it expect the **TOP LEFT** vertex to go first on each face. If any other vertex is first, it will assign the wrong light values, because it always provides top-left first on the list of light values!
So, the problem isn't that they show in the wrong face, the problem is t hat with your model, they show 180 degrees rotated! With other models it could be 0, 90, or -90. 
The solution is going to be either to teach the lighting engine to actually look at where each vertex is, instead of blindly assigning light values, or teach the model loading system to just cycle around the vertices to make them "top-left first".
I think I'll attempt the latter first.

**Comment by StanCEmpire** [comment]
Just a question regarding this issue. Would the fix for this issue require OBJ models to be modelled differently within 3D softwares or would it internally direct the lighting engine to the correct vertices and assign their correct light values light values.

In other words would teaching the lighting engine to assign the correct light values to vertices (instead of "blindly assigning light values") mean that modders would need to change how they make models.

 It seems like you're indicating that it can be done internally without modders having to edit their models to suit this. Correct me if im wrong, however. Just want to know in advance.

**Comment by TeamSpen210** [comment]
This would ensure that you do not need to do anything to your models.

**Comment by StanCEmpire** [comment]
I figured that would be the case. I guess unless something changes is the way the actual model is being loaded by forge then modelling should remain the same. Cheers for the speedy reply mate :)

**Comment by StanCEmpire** [comment]
I haven't played around with OBJ models for a bit and I'm planning on using them in the mod I'm currently working on. Just curious what progress has been made on this issue so far. If none, thats totally fine just curious where we are up to with this issue.

**Comment by StanCEmpire** [comment]
Ok so I think this issue may be related but i thought I would mention it anyway. So in forge build 31.1.1 I am having this issue where now the shadows arent appearing in random spots, but rather, my model is "immune" (so to speak) of shadows. This is the same block OBJ model used in the previous examples


Images display player inside a room which should be completely dark but the roof with my OBJ block is not dark
![2](https://user-images.githubusercontent.com/59638278/74121739-ddab2e80-4c1c-11ea-97a9-734e17f07e09.png)
![1](https://user-images.githubusercontent.com/59638278/74121743-e3087900-4c1c-11ea-872e-05c566d257f5.png)

Json Model File

```json
{

	"parent": "forge:block/default",
	"loader": "forge:obj",
	"model": "stancempiresmultidragoneggmod:models/block/test_block.obj",
	"textures":
	{
	
		"particle": "stancempiresmultidragoneggmod:block/test_block"
	
	},
	"flip-v": true

}```

**Comment by StanCEmpire** [comment]
Bump 👊

**Comment by StanCEmpire** [comment]
Ok so I've done a bit more testing [Forge Build: 31.1.16] with a couple of different models with different shapes (so some non-cube examples) and I think the issue with blocks being dark (the original concern) is fixed. I think the only problem now the blocks not having shadows and not reacting properly to varied light values (as stated above).

Im still not sure if it is a related issue to the vertex ordering issue or if it is something completely separate. I'll let someone else be the judge of that though. I just wanted to confirm that yes i have done some tests with some non-cube shapes and the issue does persist in the same way.

**Comment by StanCEmpire** [comment]
Issue is still present in 31.1.27

**Comment by StanCEmpire** [comment]
Changes in 31.1.39 have not resolved the issue. Vertex ordering still appears to be an issue, and blocks are still not reacting to light very well.

Well fully enclosed in the block, the ground (grass block) reacts to the light, but the custom OBJ blocks are not
![hmf](https://user-images.githubusercontent.com/59638278/78856936-630b5d80-7a6b-11ea-958d-6a5d6c0c5419.png)

Verticies being in the wrong order causing the shadows to appear on the upper side on the block
![hmff](https://user-images.githubusercontent.com/59638278/78856987-82a28600-7a6b-11ea-9d6e-1347d232f1d8.png)

However I do appreciate the work that is being done towards fixing this.

**Comment by tterrag1098** [comment]
Did you enable the forge pipeline?

**Comment by StanCEmpire** [comment]
I don't think so sorry, how would I go about doing that

**Comment by gigaherz** [comment]
If you look at the forge config, you should find a `experimentalForgeLightingPipelineEnabled` setting.

**Comment by StanCEmpire** [comment]
Ok cheers mate, i'll take a look into it and test again

**Comment by StanCEmpire** [comment]
Well, I can confirm that the experimental forge pipeline does indeed work

![1](https://user-images.githubusercontent.com/59638278/78874072-25b8c700-7a8f-11ea-8fd7-aeb70e76ac65.png)
![2](https://user-images.githubusercontent.com/59638278/78874083-294c4e00-7a8f-11ea-8206-6b7a40d39862.png)

Vertex ordering is correct, and lighting is working correctly with blocks

Unless there are any further concerns I think we can safely close this issue and mark as fixed

**Comment by tterrag1098** [comment]
Leaving this open until the pipeline is finalized and enabled by default.

**Comment by StanCEmpire** [comment]
Ok that makes sense, sorry for closing early

**Comment by ShetiPhian** [comment]
The experimental pipeline fixes obj models but I have a case where it breaks a json model.

![render](https://user-images.githubusercontent.com/1643122/87149873-09d7e380-c27f-11ea-9dba-6e44aaff4ece.png)

Also tested with and without 'forgeLightPipelineEnabled' but didn't seem to make any difference.

**Comment by StanCEmpire** [comment]
Just wondering at what stage is this issue. Has the pipeline been fully implemented by default or is there still work to be done? Just curious as to what stage of development this is at, and whether this issue can be closed yet.

**Comment by ShetiPhian** [comment]
Doesn't appear to be fully implemented yet.
While the experimental pipeline looks to be fully fixing OBJ models, the config still has it disabled by default and referred to as experimental. Also I'm still seeing broken JSON models on the latest builds with the option enabled. (pictured above)


I should have posted this back in my previous post but here is a json model that doesn't render correctly with the experimental pipeline: https://gist.githubusercontent.com/ShetiPhian/da2644fd23b19b3df4da587e3310fc4a/raw/5ce8c13b77f1c7def8a4fd72f782ddd2b94edff7/example.json
Its based on the model pictured above, just merged the parts into a single model and changed the textures to white concrete so additional files where not needed.

**Comment by Hulifier** [comment]
Sorry if this is common knowledge about obj models and forge. I think that OBJ models being full bright or almost full bright in the dark is caused by the "Ka" value (ambient color) in the mtl file and the "ambientToFullbright" value (only for the obj loader) in the json model being true(which is the default value).

In short, the mtl Ka value defines the texture's light 'emission', 0.0 is a normal texture without minimum light. The experimental pipeline in 1.15.2 disables this feature, but it already works in 1.16.4.

Ka 0.4 0.4 0.4
![2021-01-18_02 28 11](https://user-images.githubusercontent.com/77600669/104875931-445cc000-5935-11eb-84a8-976e4e2096d3.png)

Ka 1.0 1.0 1.0
![2021-01-18_02 29 07](https://user-images.githubusercontent.com/77600669/104875963-59395380-5935-11eb-87cc-854f978fc418.png)

the model vertex order apparently still a problem.

![2021-01-18_03 01 18](https://user-images.githubusercontent.com/77600669/104877837-7cfe9880-5939-11eb-8c23-9d426f36ee98.png)
